python/private/py_wheel.bzl

```diff
@@ -17,6 +17,62 @@
 load("//python/private:stamp.bzl", "is_stamping_enabled")
 load(":py_package.bzl", "py_package_lib")
 
+RequirementsInfo = provider(
+    "Information about all the requirements of a py_wheel.",
+    fields = {
+        "requirements": "The list of requirements.",
+    },
+)
+
+def _requirements_collector_impl(_, ctx):
+    """Aspect to collect third-party dependencies.
+
+    Args:
+        _: The bazel target the aspect is applied to.
+        ctx: The aspect context.
+    Returns: RequirementsInfo
+    """
+
+    requirement = None
+    if hasattr(ctx.rule.attr, "tags"):
+        pypi_name = None
+        pypi_version = None
+        for tag in ctx.rule.attr.tags:
+            if tag.startswith("pypi_name="):
+                pypi_name = tag[len("pypi_name="):]
+            elif tag.startswith("pypi_version="):
+                pypi_version = tag[len("pypi_version="):]
+
+        if pypi_name and pypi_version:
+            requirement = "{}=={}".format(pypi_name, pypi_version)
```

Comment:
You already know: this is troubling, that ppl will publish libraries to pypi with exact version pinning like this rather than only a lower-bound. It will lead to incompatibilities downstream, when a constraint solver can't find a solution.

I suppose the alternative is for us to lookup the constraint from the `requirements.in` file (but could be a constraints.txt too).

Or maybe we should let users choose to use some semver range operator other than `==` in the thing we generate here?

Reply:
Addressed in f2d7ae1. It can be a later work if we see the need to be fancier.